HostBuilder.Build doc: mention lifetime

yecril71pl opened the pull request:
Mention that the host application lifetime service will be added *in addition* to the services configured to be run.

yecril71pl committed:
HostBuilder.Build doc: mention lifetime

Mention that the host application lifetime service will be added *in addition* to the services configured to be run.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -119,6 +119,8 @@ public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderCont
         /// <summary>
         /// Run the given actions to initialize the host. This can only be called once.
+        /// Adds host application lifetime service to the host.
+        /// The application will run until interrupted or until <see cref="M:IHostApplicationLifetime.StopApplication()" /> is called.
```
```suggestion
```

I don't think this line makes sense on the `Build()` method.

yecril71pl replied:
I would very much appreciate having this explained somewhere though.  I have put it here because this is what causes this behaviour.  Where would you expect it to appear instead?

eerhardt replied:
> I have put it here because this is what causes this behaviour. Where would you expect it to appear instead?

I would expect it on `Start` or `Run`. If you just call `Build()` without calling `Start`, the application won't run until interrupted.

yecril71pl replied:
But you can also explicitly remove the lifetime service using [RemoveAll](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.removeall#Microsoft_Extensions_DependencyInjection_Extensions_ServiceCollectionDescriptorExtensions_RemoveAll__1_Microsoft_Extensions_DependencyInjection_IServiceCollection_) and if you do that, it will not keep the application running, so the documentation for [StartAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihost.startasync#Microsoft_Extensions_Hosting_IHost_StartAsync_System_Threading_CancellationToken_) would have to say *The default implementation also includes running the [ApplicationLifetime](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.internal.applicationlifetime) that blah blah blah unless you have explicitly removed it in [Build](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.hostbuilder.build#Microsoft_Extensions_Hosting_HostBuilder_Build)*, which is possible but sounds weird without having it documented on [Build](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.hostbuilder.build#Microsoft_Extensions_Hosting_HostBuilder_Build) itself.

yecril71pl replied:
OK.  I guess a remark at [StartAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihost.startasync#Microsoft_Extensions_Hosting_IHost_StartAsync_System_Threading_CancellationToken_) to that effect would be appropriate too?  Because it does not make sense to mention the service if it is not required.

eerhardt replied:
I'm not sure I understand the question. I don't think we need to document that if you remove all the Lifetime services it will throw, as I don't see that as a common scenario for someone to do.

yecril71pl replied:
I can move the mention of the lifetime service to [StartAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihost.startasync#Microsoft_Extensions_Hosting_IHost_StartAsync_System_Threading_CancellationToken_) and if you had removed it, you would get an undocumented exception at run time.  Does that sound OK?

eerhardt replied:
I don't think we need to document that if you remove all the Lifetime services it will throw, as I don't see that as a common scenario for someone to do.

I think it would be reasonable to add to the StartAsync document that "The application will run until interrupted or until <see cref="M:IHostApplicationLifetime.StopApplication()" /> is called." Today it just says "Starts the program." which isn't that helpful.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -119,6 +119,8 @@ public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderCont
 
         /// <summary>
         /// Run the given actions to initialize the host. This can only be called once.
+        /// Adds host application lifetime service to the host.
```
It adds other services as well, for example logging, configuration, etc. Why are the lifetime services special here?

If we want to list the services added here, I think it would be more appropriate to add them to a `<remarks>` section instead.

yecril71pl replied:
Lifetime services are special because they prevent the application from quitting and you must explicitly ask the application to quit from client code, whereas all other services are used as needed and do not affect the main workflow.  That is, when you tell the application to run as a background service, nothing tells you that it will run indefinitely and this is not what you, as a whatever-else programmer, would expect (a whatever-else programmer would expect an explicit call to `EnterMainEventLoop` or whatever).  I tell you this as a whatever-else programmer.  Surprise factor 90%.

eerhardt replied:
The current worker template has the following code in Program.cs:

```C#
using WorkerService6;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddHostedService<Worker>();
    })
    .Build();

await host.RunAsync();
```

The equivalent place to `EnterMainEventLoop` is the last line: `await host.RunAsync();`. This is the place where your program will run indefinitely until something cancels it.

yecril71pl replied:
Yes, and this is unexpected because the code in the `Worker` provides no indication of an infinite loop and the callback method in `Worker` is called only once.  This is why it looks odd at the first encounter.

eerhardt replied:
The template has a loop in it:

```C#
namespace WorkerService6
{
    public class Worker : BackgroundService
    {
        private readonly ILogger<Worker> _logger;

        public Worker(ILogger<Worker> logger)
        {
            _logger = logger;
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);
                await Task.Delay(1000, stoppingToken);
            }
        }
    }
}
```

The idea being that this background task loops until it is done with its work.

Check out https://docs.microsoft.com/en-us/dotnet/core/extensions/workers, which describes how all the pieces fit together.

eerhardt replied:
I think it would be fine to add a general statement to the `<remarks>` section that said something like:

```
Adds basic services to the host such as application lifetime, host environment, and logging.
```

yecril71pl replied:
> The template has a loop in it:

I did not use the template because I did not know it exists; the documentation for [BackgroundService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.backgroundservice) does not mention it at all.  Do you think it should?  I just inherited from [BackgroundService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.backgroundservice), implemented a straightforward procedure and was stupefied by the effect.

eerhardt replied:
I think it would be reasonable for the documentation for BackgroundService to point to https://docs.microsoft.com/dotnet/core/extensions/workers.

@gewarren @IEvangelist - do we have any examples of the xml docs linking to the conceptual docs? Here's a case where I think it would be helpful.

IEvangelist replied:
I'm not actually sure, but I do agree that would be helpful. Ideally, it would exclude the locale as you've done there. Checkout this [grep.app search results](https://grep.app/search?q=docs.microsoft.com/en-us/dotnet&filter[repo][0]=dotnet/runtime).

gewarren replied:
You can add a `<related>` tag like [this](https://github.com/dotnet/runtime/blob/b2ab353c0ed8eb9d394570a7cde8c4ed03c771e5/src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs#L691), which adds a link under the **See also** heading.

yecril71pl committed:
ExecuteAsync: See Worker Services in .NET

yecril71pl committed:
StartAsync: mention IHostApplicationLifetime

1. The application will run until interrupted or until `HostApplicationLifetime.StopApplication()` is called.
2. Document how to configure `Services`.

yecril71pl committed:
mv mention StopApplication to StartAsync

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

yecril71pl committed:
Build doc: mv services to remarks

yecril71pl committed:
typo fix

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHost.cs:
```diff
@@ -13,12 +13,13 @@ namespace Microsoft.Extensions.Hosting
 
         /// <summary>
-        /// Start the program.
+        /// Start the <see cref="P:IHost.Services" /> configured for the program.
```
```suggestion
        /// Start the <see cref="IHostedService" /> objects configured for the program.
```

I wouldn't reference the `Services` property here. Those are all the DI services for the program. They don't get started by this. Instead, the `IHostedService` instances do.

eerhardt reviewed (commented):
This looks great @yecril71pl! Thanks for the contribution, and your patience with my nits.

I just had one small additional comment, and I think this will be ready to merge.

yecril71pl committed:
Starts only hosted services

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

yecril71pl commented:
> This looks great @yecril71pl! Thanks for the contribution, and your patience with my nits.
> 
> I just had one small additional comment, and I think this will be ready to merge.

Thank you for helping me to understand how these things work.  It all looks like black magic when you see it for the first time.

eerhardt reviewed (approved):
LGTM. Thanks again!

gewarren commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -28,6 +28,7 @@ public abstract class BackgroundService : IHostedService, IDisposable
         /// the lifetime of the long running operation(s) being performed.
         /// </summary>
         /// <param name="stoppingToken">Triggered when <see cref="IHostedService.StopAsync(CancellationToken)"/> is called.</param>
+        /// See <related type="Article" href="/dotnet/core/extensions/workers">Worker Services in .NET</related> for implementation guidelines.
```
This is incorrect, AFAIK. The `<related>` tag shouldn't have any surrounding text. If you want to have descriptive text like that, I think you need to use a `<see href="url">text</see>` within the summary or remarks.

gewarren commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHost.cs:
```diff
@@ -13,12 +13,13 @@ namespace Microsoft.Extensions.Hosting
     {
         /// <summary>
-        /// The programs configured services.
+        /// The services configured for the program (e.g. using <see cref="M:HostBuilder.ConfigureServices(Action&lt;HostBuilderContext,IServiceCollection&gt;)" />).
```
```suggestion
        /// Gets the services configured for the program (for example, using <see cref="M:HostBuilder.ConfigureServices(Action&lt;HostBuilderContext,IServiceCollection&gt;)" />).
```

gewarren commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHost.cs:
```diff
@@ -13,12 +13,13 @@ namespace Microsoft.Extensions.Hosting
 
         /// <summary>
-        /// Start the program.
+        /// Start the <see cref="IHostedService" /> objects configured for the program.
```
```suggestion
        /// Starts the <see cref="IHostedService" /> objects configured for the program.
```

gewarren commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -121,6 +121,7 @@ public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderCont
         /// Run the given actions to initialize the host. This can only be called once.
         /// </summary>
         /// <returns>An initialized <see cref="IHost"/></returns>
+        /// <remarks>Adds basic services to the host such as application lifetime, host environment and logging.</remarks>
```
Microsoft uses Oxford commas.

```suggestion
        /// <remarks>Adds basic services to the host such as application lifetime, host environment, and logging.</remarks>
```

yecril71pl committed:
See see

yecril71pl committed:
IHost.Services are RO

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

yecril71pl committed:
Use indicative

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

yecril71pl committed:
Oxford commas

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -28,6 +28,7 @@ public abstract class BackgroundService : IHostedService, IDisposable
         /// the lifetime of the long running operation(s) being performed.
         /// </summary>
         /// <param name="stoppingToken">Triggered when <see cref="IHostedService.StopAsync(CancellationToken)"/> is called.</param>
+        /// See <see href="https://docs.microsoft.com/dotnet/core/extensions/workers">Worker Services in .NET</see> for implementation guidelines.
```
```suggestion
        /// <related type="Article" href="/dotnet/core/extensions/workers">Worker Services in .NET</related>
```

To follow the example @gewarren linked to: https://github.com/dotnet/runtime/blob/b2ab353c0ed8eb9d394570a7cde8c4ed03c771e5/src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs#L691

yecril71pl replied:
But that is a bare reference, with no information why it is there.  I would prefer to have it in context.  Please confirm.

gewarren replied:
If you decide to keep the `<see>` tag, you'll need to put the entire text within the `summary` or `remarks` tags.

eerhardt replied:
The docs understand this format, and it makes it look like this:

https://docs.microsoft.com/en-us/dotnet/api/system.numerics.vector2.tostring?view=net-6.0#System_Numerics_Vector2_ToString_System_String_

![image](https://user-images.githubusercontent.com/8291187/148264207-5a5f6be9-9f63-41f4-a84d-99a421d49ef6.png)

yecril71pl replied:
Yes, and that is for non-essential references, like when you can safely assume the reader already knows these things.  This is not the case here.

yecril71pl committed:
see in remarks

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -28,8 +28,9 @@ public abstract class BackgroundService : IHostedService, IDisposable
-        /// See <see href="https://docs.microsoft.com/dotnet/core/extensions/workers">Worker Services in .NET</see> for implementation guidelines.
         /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        /// <remarks>See <see href="https://docs.microsoft.com/dotnet/core/extensions/workers">Worker Services in .NET</see> for implementation guidelines.</remarks>
+
```
```suggestion
```

yecril71pl committed:
typo fix

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>
